pkg/helpers/identifier_string_parser_test.go

```diff
@@ -91,4 +94,251 @@ func Test_ParseIdentifierString(t *testing.T) {
+		Expected sdk.ExternalObjectIdentifier
+		Error    string
+	}{
+		{Input: ``, Error: "incompatible identifier: "},
+		{Input: "a\nb.cde.efg", Error: "unable to read identifier: a\nb.cde.efg, err = record on line 2: wrong number of fields"},
+		{Input: `a"b.cde.efg`, Error: "unable to read identifier: a\"b.cde.efg, err = parse error on line 1, column 2: bare \" in non-quoted-field"},
+		{Input: `abc.cde.efg.ghi`, Error: `unexpected number of parts 4 in identifier abc.cde.efg.ghi, expected 3 in a form of "<organization_name>.<account_name>.<external_object_name>"`},
+		{Input: `abc.cde`, Error: `unexpected number of parts 2 in identifier abc.cde, expected 3 in a form of "<organization_name>.<account_name>.<external_object_name>"`},
+		{Input: `""."".""`, Expected: sdk.NewExternalObjectIdentifier(sdk.NewAccountIdentifier(``, ``), sdk.NewAccountObjectIdentifier(``))},
+		{Input: `""""."""".""""`, Expected: sdk.NewExternalObjectIdentifier(sdk.NewAccountIdentifier(`"`, `"`), sdk.NewAccountObjectIdentifier(`"`))},
+		{Input: `abc.cde.efg`, Expected: sdk.NewExternalObjectIdentifier(sdk.NewAccountIdentifier(`abc`, `cde`), sdk.NewAccountObjectIdentifier(`efg`))},
+		{Input: `"abc"."cde"."efg"`, Expected: sdk.NewExternalObjectIdentifier(sdk.NewAccountIdentifier(`abc`, `cde`), sdk.NewAccountObjectIdentifier(`efg`))},
+		{Input: `"ab.c"."cd.e"."ef.g"`, Expected: sdk.NewExternalObjectIdentifier(sdk.NewAccountIdentifier(`ab.c`, `cd.e`), sdk.NewAccountObjectIdentifier(`ef.g`))},
+		{Input: `"a""bc"."cd""e"."ef""g"`, Expected: sdk.NewExternalObjectIdentifier(sdk.NewAccountIdentifier(`a"bc`, `cd"e`), sdk.NewAccountObjectIdentifier(`ef"g`))},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(fmt.Sprintf(`Parsing external object identifier with input: "%s"`, testCase.Input), func(t *testing.T) {
+			id, err := ParseExternalObjectIdentifier(testCase.Input)
+
+			if testCase.Error != "" {
+				assert.ErrorContains(t, err, testCase.Error)
+			} else {
+				assert.Equal(t, testCase.Expected, id)
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func Test_Encoding_And_Parsing_Of_ResourceIdentifier(t *testing.T) {
+	testCases := []struct {
+		Input                 []string
+		Expected              string
+		ExpectedAfterDecoding []string
+	}{
+		{Input: []string{sdk.NewSchemaObjectIdentifier("a", "b", "c").FullyQualifiedName(), "info"}, Expected: `"a"."b"."c"|info`},
+		{Input: []string{}, Expected: ``},
+		{Input: []string{"", "", ""}, Expected: `||`},
+		{Input: []string{"a", "b", "c"}, Expected: `a|b|c`},
+		// If one of the parts contains a separator sign (pipe in this case),
+		// we can end up with more parts than we started with.
+		{Input: []string{"a", "b", "c|d"}, Expected: `a|b|c|d`, ExpectedAfterDecoding: []string{"a", "b", "c", "d"}},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(fmt.Sprintf(`Encoding and parsing %s resource identifier`, testCase.Input), func(t *testing.T) {
+			encodedIdentifier := EncodeResourceIdentifier(testCase.Input...)
```

Comment:
How about we use real sdk.*Identifier types instead of []string? Potentially, we could have helpers to extract `parts` for these types so they can be easily used here.

Reply:
That made me think about how we actually want to save our IDs in the state. I think for IDs represented by only Snowflake ID we could do just `d.SetId(id.FullyQualifiedName())` and `id, err := ParseXIdentifier(d.Id)`. For more complex identifiers (like grant ones) we could use this one as `EncodeResourceIdentifier(id.FullyQualifiedName(), "Type", "some other data")` and `parts := ParseResourceIdentifier; do the stuff with parts like id parsing from the first part`. With that in mind, there's no need for proposed logic (or maybe I misunderstood it).